src/xrpld/app/tx/detail/Batch.cpp

```diff
@@ -0,0 +1,275 @@
+    Copyright (c) 2024 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <xrpld/app/tx/apply.h>
+#include <xrpld/app/tx/detail/Batch.h>
+#include <xrpld/ledger/Sandbox.h>
+#include <xrpld/ledger/View.h>
+#include <xrpl/basics/Log.h>
+#include <xrpl/protocol/Feature.h>
+#include <xrpl/protocol/Indexes.h>
+#include <xrpl/protocol/TER.h>
+#include <xrpl/protocol/TxFlags.h>
+
+namespace ripple {
+
+TxConsequences
+Batch::makeTxConsequences(PreflightContext const& ctx)
+{
+    return TxConsequences{ctx.tx, TxConsequences::normal};
+}
+
+NotTEC
+Batch::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureBatch))
+        return temDISABLED;
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    auto& tx = ctx.tx;
+
+    if (tx.getFlags() & tfBatchMask)
+    {
+        JLOG(ctx.j.warn()) << "Batch: invalid flags.";
+        return temINVALID_FLAG;
+    }
+
+    AccountID const outerAccount = tx.getAccountID(sfAccount);
+
+    auto const& txns = tx.getFieldArray(sfRawTransactions);
+    STVector256 const hashes = tx.getFieldV256(sfTxIDs);
+    if (hashes.size() != txns.size())
+    {
+        JLOG(ctx.j.warn()) << "Batch: Hashes array size does not match txns.";
+        return temINVALID;
+    }
+
+    if (txns.empty())
+    {
+        JLOG(ctx.j.warn()) << "Batch: txns array empty.";
+        return temARRAY_EMPTY;
+    }
+
+    if (txns.size() > 8)
+    {
+        JLOG(ctx.j.warn()) << "Batch: txns array exceeds 12 entries.";
+        return temARRAY_TOO_LARGE;
+    }
+
+    for (int i = 0; i < txns.size(); ++i)
+    {
+        STTx const stx = STTx{STObject(txns[i])};
+        if (stx.getTransactionID() != hashes[i])
+        {
+            JLOG(ctx.j.warn()) << "Batch: Hashes array does not match txns.";
+            return temINVALID;
+        }
+
+        if (!stx.isFieldPresent(sfTransactionType))
+        {
+            JLOG(ctx.j.warn())
+                << "Batch: TransactionType missing in array entry.";
+            return temINVALID_BATCH;
+        }
+        if (stx.getFieldU16(sfTransactionType) == ttBATCH)
```

Comment:
are there additional transactions that we want to restrict from being an inner transaction?  there are some new tx like `LedgerStateFix`. cc @mvadari

Reply:
I think `Batch` is probably the only one we want to restrict, to prevent nesting - I don't really see a reason to _prevent_ someone from batching with a `LedgerStateFix` tx (though I do think it'd be weird to do that).

Reply:
Cool, yeah I think we'd need keep an eye on future new transactions that we'd want to consider preventing from being an inner tx.